src/sinks/connections/TcpClient.ts

```diff
@@ -24,13 +24,23 @@ export class TcpClient implements ISocketClient {
 
   constructor(endpoint: IEndpoint) {
     this.endpoint = endpoint;
-    this.socket = new net.Socket({ allowHalfOpen: true, writable: false })
+    this.socket = new net.Socket({ allowHalfOpen: true })
       .setEncoding('utf8')
       .setKeepAlive(true)
       .setTimeout(5000) // idle timeout
       .on('timeout', () => this.disconnect('idle timeout'))
       .on('end', () => this.disconnect('end'))
       .on('data', data => LOG('TcpClient received data.', data));
+    this.init.apply(this);
+  }
+
+  public async init(): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      this.socket.connect(this.endpoint.port, this.endpoint.host, (err?: Error) => {
```

Comment:
So, if I understand this correctly, by doing an early connect attempt here we cover the case where the socket was already connected. We just throw so that the next connect can work? How exactly will this guarantee that `close` is emitted before we try to connect next?

Reply:
We need to get rid of the `writable: false` socket property which was causing the socket to be in `readOnly` mode for newer node versions, but doing that brings back [this](https://github.com/awslabs/aws-embedded-metrics-node/commit/0034ce1cd5e7109a43f3a1551a985ffa984247ce) issue. To avoid that we need to connect to the socket after creation as recommended by the node docs.

There are 5 states the socket can be in: `open`, `opening`, `readOnly`, `writeOnly` and `closed`.
- The `open` and `opening` states are already being [handled](https://github.com/awslabs/aws-embedded-metrics-node/blob/master/src/sinks/connections/TcpClient.ts#L103-L112) by the code.
- The `readOnly` and `writeOnly` states won't happen since we aren't setting the socket to `readable` or `writable` anywhere.
- The `closed` state calls `connect()` or on error calls `disconnect()` and then tries to reconnect.